virtool/blast/task.py

```diff
@@ -53,45 +53,69 @@ async def request(self):
 
-    async def wait(self):
+    async def wait(self, retries: int = 3):
         """
         Wait until the BLAST search completes.
 
         Keep check the BLAST status on NCBI with increasingly longer intervals between
         checks.
 
-        Checks are conducted by the data layer and will store the results or error when
-        the search completes. The task completes when either an error or
+        Checks are conducted by the data layer and will store the results or error
+        when the search completes. The task completes when either an error or
 
+        The BLAST will be retried up to 3 times if a single BLAST
+        search exceeds 10 minutes
         """
-        try:
-            while True:
-                await asyncio.sleep(self.interval)
 
-                blast = await self.data.blast.check_nuvs_blast(
+        async def wait_with_timeout():
+            try:
+                while True:
+                    await asyncio.sleep(self.interval)
+
+                    blast = await self.data.blast.check_nuvs_blast(
+                        self.analysis_id, self.sequence_index
+                    )
+
+                    if blast.ready:
+                        logger.info("Retrieved result for BLAST %s", blast.rid)
+                        break
+
+                    if blast.error:
+                        logger.info(
+                            "Encountered error for BLAST %s: %s",
+                            blast.rid,
+                            blast.error,
+                        )
+                        await self._set_error(blast.error)
+                        break
+
+                    self.interval += 5
+
+                    logger.debug(
+                        "Checked BLAST %s. Waiting %s seconds",
+                        blast.rid,
+                        self.interval,
+                    )
+
+            except asyncio.CancelledError:
+                await self.data.blast.delete_nuvs_blast(
                     self.analysis_id, self.sequence_index
                 )
 
-                if blast.ready:
-                    logger.info("Retrieved result for BLAST %s", blast.rid)
-                    break
+                logger.info("Deleted BLAST due to cancellation: %s", self.rid)
 
-                if blast.error:
-                    logger.info(
-                        "Encountered error for BLAST %s: %s", blast.rid, blast.error
-                    )
-                    await self._set_error(blast.error)
-                    break
+        retries: int = 0
 
-                self.interval += 5
+        while True:
+            try:
+                await asyncio.wait_for(wait_with_timeout(), 10)
+                break
 
-                logger.debug(
-                    "Checked BLAST %s. Waiting %s seconds", blast.rid, self.interval
-                )
+            except asyncio.TimeoutError:
+                if retries < 3:
+                    retries += 1
```

Comment:
Stylistic thing, but I personally check fo `retries >= 3` then bail out if that is the case. Allows continuing to be the default case